bundles/org.platformio.eclipse.ide.installer/src/org/platformio/eclipse/ide/installer/base/BaseEnvironment.java

```diff
@@ -0,0 +1,116 @@
+/*******************************************************************************
+ * Copyright (c) 2021 ArSysOp and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0/.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Nikifor Fedorov (ArSysOp) - initial API and implementation
+ *******************************************************************************/
+package org.platformio.eclipse.ide.installer.base;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.platformio.eclipse.ide.installer.api.CommandResult;
+import org.platformio.eclipse.ide.installer.api.Environment;
+import org.platformio.eclipse.ide.installer.api.OS;
+
+public final class BaseEnvironment implements Environment {
+
+	private final OS os;
+	private final Map<String, Process> running = new HashMap<>();
+
+	public BaseEnvironment(OS os) {
+		this.os = os;
+	}
+
+	private List<String> input(String command, List<String> args) {
+		List<String> resolved = new LinkedList<>();
+		resolved.add(command);
+		resolved.addAll(args);
+		return resolved;
+	}
+
+	@Override
+	public CommandResult execute(String command, List<String> arguments) {
+		ProcessBuilder builder = new ProcessBuilder(input(command, arguments));

```

Comment:
You need to employ stream readers from other threads here. Otherwise the process may never end